### ISSUE TITLE
Do we need to optimize the method `c.ObjectWatcher.NeedsUpdate`?

### ISSUE THREAD

**Issue by yike21** [karmada-io/karmada#3564]
**Please provide an in-depth description of the question you have**:
When I bundle third-party resources into the Resource Interpreter framework and write ResourceInterpreterCustomization for `argoproj.io/v1alpha1/Workflow`, I find that a reconcile loop happened between `work-status-controller` and `workflow-controller`.

The known execution information of `workflow` is as follows:
`workflow` needs to run in phases, and for each completed phase, the `.metadata.generation` of the workflow will increase and the `.status.nodes` will be updated with information on the current phase.

Let's imagine a scenario：
Suppose a **workload** such as this:
1. There exists a workload that can release pods in batches, and the `.metadata.generation` changes for each batch released.
2. The workload is marked as completed by `.status.phase==completed` when all batches are successful.
3. This workload runs in multiple batches, and before releasing the next batch of pods, it will re-confirm that the current batch of pods has been successfully released. We suppose that the confirmation operation relies on the workload's `.status.nodes`.
4. If for some reason the workload loses the `.status.nodes` information when a batch is about to be published, the workload needs to start again from batch 1 since the confirmation operation fails.

Then I propagate this workload to a member cluster:
1. When this workload in member cluster publishes the first batch of pods, its `.metadata.generation` is increased and `.status.nodes` records the information.
2. `work-status-controller` needs to [syncWorkStatus](https://github.com/karmada-io/karmada/blob/master/pkg/controllers/status/work_status_controller.go#L167) because the status of the workloads in the member cluster has changed  (*here I think work-status-controller will reconcile work*). At this point, [NeedsUpdate](https://github.com/karmada-io/karmada/blob/master/pkg/controllers/status/work_status_controller.go#L218) returns true because the `.metadata.generation` of workload in member cluster has increased and is different from that in `workObject.spec.workload.manifests`.
3. At [c.ObjectWatcher.Update(clusterName, desiredObj, observedObj)](https://github.com/karmada-io/karmada/blob/master/pkg/controllers/status/work_status_controller.go#L224), `work-status-controller` will use the workload stored in `workObject.spec.workload.manifests` that does not contain `.status.nodes` to update the workloads in member cluster.
4. The workload in member cluster loses `.status.nodes` information, and the workload cannot continue to publish the second batch of pods and must start again from the first batch.

In this case, a loop is formed around `.metadata.generation` and `.status.nodes` between the `work-status-controller` of karmada-control-plane and the `workload-controller` of the member cluster.

**What do you think about this question?**:
This problem can also be described as “How to handle CRDs that run in phases and are dependent on the previous phase?”

A possible way to handle it is to retain the entire `.status` of CRD like [this](https://github.com/karmada-io/karmada/pull/3438#issuecomment-1549521651), [related code](https://github.com/karmada-io/karmada/pull/3438/files#diff-9c0240abf3714a93e53c985e0135abedec49a0bdbe0a3f496f4f79fc7d011a47R297-R299). By this way, `workObject.spec.workload.manifests` has the `.status` of workload. After [c.ObjectWatcher.Update(clusterName, desiredObj, observedObj)](https://github.com/karmada-io/karmada/blob/master/pkg/controllers/status/work_status_controller.go#L224), the workload in member cluster retain`.status.nodes` information.

Or we may optimize the method `c.ObjectWatcher.NeedsUpdate` in some way to avoid updating `workload` before`workload` in member cluster is not completed.

**Environment**:
- Karmada version: latest
- Kubernetes version:
- Others:

**Comment by XiShanYongYe-Chang** [comment]
Thanks for the detailed description.

> * At [c.ObjectWatcher.Update](https://github.com/karmada-io/karmada/blob/master/pkg/controllers/status/work_status_controller.go#L224), `work-status-controller` will use the workload stored in `workObject.spec.workload.manifests` that does not contain `.status.nodes` to update the workloads in member cluster.
> * The workload in member cluster loses `.status.nodes` information, and the workload cannot continue to publish the second batch of pods and must start again from the first batch.

There is a question: When we are updating the resources of the resource, the status part should not be updated. Why does the status in the member cluster lose?

**Comment by chaunceyjiang** [comment]
> When we are updating the resources of the resource, the status part should not be updated. Why does the status in the member cluster lose?

Because the `.status` of workflow is not a subresource.

**Comment by chaunceyjiang** [comment]
> possible way to handle it is to retain the entire .status of CRD like https://github.com/karmada-io/karmada/pull/3438#issuecomment-1549521651, [related code](https://github.com/karmada-io/karmada/pull/3438/files#diff-9c0240abf3714a93e53c985e0135abedec49a0bdbe0a3f496f4f79fc7d011a47R297-R299).


I think this is a good way.

**Comment by XiShanYongYe-Chang** [comment]
> Because the `.status` of workflow is not a subresource.

Oh. thx, I got it.

**Comment by XiShanYongYe-Chang** [comment]
> > possible way to handle it is to retain the entire .status of CRD like [#3438 (comment)](https://github.com/karmada-io/karmada/pull/3438#issuecomment-1549521651), [related code](https://github.com/karmada-io/karmada/pull/3438/files#diff-9c0240abf3714a93e53c985e0135abedec49a0bdbe0a3f496f4f79fc7d011a47R297-R299).
> 
> I think this is a good way.

Dose this method should directly modify the lua script, if so, I think this modification is fine.

**Comment by yike21** [comment]
> > > possible way to handle it is to retain the entire .status of CRD like [#3438 (comment)](https://github.com/karmada-io/karmada/pull/3438#issuecomment-1549521651), [related code](https://github.com/karmada-io/karmada/pull/3438/files#diff-9c0240abf3714a93e53c985e0135abedec49a0bdbe0a3f496f4f79fc7d011a47R297-R299).
> > 
> > 
> > I think this is a good way.
> 
> Dose this method should directly modify the lua script, if so, I think this modification is fine.

Yes, It just modify the lua script when bundling third-part  resources.  

And I have an oversight: the premise is that workload's .status is not a subresource, then the `status` in the member cluster lose. 
 `workflow` fits this case

**Comment by XiShanYongYe-Chang** [comment]
Thank you for your question summary.

**Comment by yike21** [comment]
Thank you for replying!
